### PR TITLE
Subtests for mojo ua

### PR DESCRIPTION
**Pull request by rabbiveesh**
### Summary
Changing the tests for Mojo::UserAgent to use subtests

### Motivation
This is a project we're working on

### References
as discussed in issue #1520 

Currently, I need someone to look this over. I have two tests which started failing and are marked as TODO, and I can't figure out what's going on.

### CODE REVIEW

**Review by kraih — CHANGES_REQUESTED** [review]
You forgot to squash commits.

**Comment by rabbiveesh** [comment]
Squashed. Could you look at the two TODO tests? I don't understand why they aren't keeping alive as expected

File: t/mojo/user_agent.t [hunk 1]
```diff
@@ -551,198 +605,214 @@ subtest 'Fork-safety' => sub {
...
+    $tx->resume;
+    return unless my $id = $tx->connection;
+    $stream += @{Mojo::IOLoop->stream($id)->subscribers('drain')};
+  };
+  $tx->req->content->$drain;
+  $ua->start($tx);
+  ok !$tx->error, 'no error';
+TODO: {
+    local $TODO = 'This one also should not have changed b/c of subtests';
+    ok $tx->kept_alive, 'kept connection alive';
```
Comment (kraih) [inline t/mojo/user_agent.t:682]: The user agent gets re-initiaslized, of course that resets the connection cache.

**Review by kraih — DISMISSED** [review]
There should be no tests in TODO blocks.

File: t/mojo/user_agent.t (hunk 1 above)
Comment (rabbiveesh) [reply t/mojo/user_agent.t:673]: I nested those two tests under the ones above them, since they depend on the previous subtest's state

Comment (kraih) [reply t/mojo/user_agent.t:673]: That's wrong. The callback tests have nothing to do with introspection.

**Review by kraih — DISMISSED** [review]
Nesting unrelated subtests is a hack, not a solution.

**Comment by rabbiveesh** [comment]
> Nesting unrelated subtests is a hack, not a solution.

Would you prefer if I removed the assertions that depend on unrelated state? Or should the `$ua` be scoped to the file, and reinstantiated when necessary?

**Comment by kraih** [comment]
Neither, the test should be fixed up so coverage stays the same and the code is clean.

**Comment by rabbiveesh** [comment]
Okay, so I switched the assertions to reflect where the ua got reinstantiated. Now there are no sub-subtests.

**Comment by jberger** [comment]
Needs a tidy.

**Review by jberger — DISMISSED** [review]

File: t/mojo/user_agent.t [hunk 2]
```diff
@@ -69,16 +69,14 @@ get '/redirect_close' => sub {
   $c->res->fix_headers->headers->remove('Content-Length');
 };
 
-# Max redirects
-{
+subtest 'Max redirects' => sub {
   local $ENV{MOJO_MAX_REDIRECTS} = 25;
   is(Mojo::UserAgent->new->max_redirects, 25, 'right value');
```
Comment (jhthorsen) [inline t/mojo/user_agent.t:74]: Is there any reason for the parenthesis here? I know it's not directly related to the PR, but I was wondering if this should be removed as well to make the syntax more consistent.

File: t/mojo/user_agent.t [hunk 3]
```diff
@@ -92,432 +90,492 @@ get '/redirect_close' => sub {
...
+  ok !$tx->kept_alive, 'kept connection not alive';
+  ok $tx->keep_alive, 'keep connection alive';
+  is $tx->res->version, '1.1', 'right version';
+  is $tx->res->code,    200,   'right status';
+  ok !$tx->res->headers->connection, 'no "Connection" value';
+  is $tx->res->max_message_size, 2147483648, 'right value';
+  is $tx->res->body,             'works!',   'right content';
+};
+
+subtest 'Again' => sub {
```
Comment (jhthorsen) [inline t/mojo/user_agent.t:236]: This is no good: I think the idea here is to make sure that "kept_alive" is true between requests. Not quite sure how to format that though. Maybe a `$ua_keep_alive` object outside of the subtests? Or maybe a "# Again" comment inside the subtest? I guess the latter is better.

File: t/mojo/user_agent.t [hunk 4]
```diff
@@ -92,432 +90,492 @@ get '/redirect_close' => sub {
...
+  $tx = $ua->start($tx);
+  ok !$tx->error, 'no error';
+  is $finished_req, 1, 'finish event has been emitted once';
+  is $finished_tx,  1, 'finish event has been emitted once';
+  is $finished_res, 1, 'finish event has been emitted once';
+  ok $tx->req->is_finished, 'request is finished';
+  ok $tx->is_finished, 'transaction is finished';
+  ok $tx->res->is_finished, 'response is finished';
+  ok !$tx->error,      'no error';
+  ok !$tx->kept_alive, 'kept connection not alive';
```
Comment (jhthorsen) [inline t/mojo/user_agent.t:366]: I'm not sure if this is acceptable. Same goes for the other kept_alive/keep_alive changes. But maybe it's ok if the "Again" test above is fixed?

File: t/mojo/user_agent.t [hunk 5]
```diff
@@ -551,198 +609,211 @@ subtest 'Fork-safety' => sub {
...
-$drain = sub {
-  my $content = shift;
-  return $ua->ioloop->timer(
-    0.25 => sub {
-      $content->write_chunk('');
-      $tx->resume;
-      $stream += @{Mojo::IOLoop->stream($tx->connection)->subscribers('drain')};
+subtest 'Introspect' => sub {
+  my $ua = Mojo::UserAgent->new(ioloop => Mojo::IOLoop->singleton);
+
```
Comment (jhthorsen) [inline t/mojo/user_agent.t:614]: Extra newline.

**Review by jhthorsen — CHANGES_REQUESTED** [review]
Using https://github.com/mojolicious/mojo/pull/1581/files?w=1 I see some test changes that I don't think is acceptable. See inline comments.

**Comment by kraih** [comment]
Since tests are actually being changed here please make sure to to an in depth review. So we can be sure coverage doesn't change.

File: t/mojo/user_agent.t (hunk 3 above)
Comment (rabbiveesh) [reply t/mojo/user_agent.t:236]: I'll tell you, I can't tell how much of these tests are actually testing an edge case, or if they were cargo culted from a previous test. So I have no idea what is affecting code coverage here or not.
All of these tests were subtly using the previous state of unrelated tests, which would lead me to believe that they were more boilerplate than real test. But I don't know. I could change it to gratuitous second requests in each of the tests that had to be changed, if that's necessary. But I don't know if it's just noise, as above.

Comment (kraih) [reply t/mojo/user_agent.t:236]: Keep-alive tests tend to be significant.

Comment (rabbiveesh) [reply t/mojo/user_agent.t:236]: so should I add an extra call wherever there was a keep-alive test? My original thought was that kept-alive mirrors keep-alive, so I can just check one of the two.

**Comment by kraih** [comment]
I'll close this PR for now, so it doesn't block someone else from converting the test.
